### DISCUSSION TITLE
Trying to run MONAI Label to autosegment heart

### DISCUSSION THREAD

**Discussion by coronaryangina001** [Project-MONAI/MONAILabel#1639] [Q&A]
Hello! I'm trying to use the segmentation pre-trained model in the radiology app to autosegment out some hearts that I plan on 3d printing! I've gotten the MonAI Label server working on my computer using Anaconda in this tutorial: https://www.youtube.com/watch?v=8y1OBQs2wis&t=154s. I've been able to upload my own data onto the server but I can't get any of the programs (deepedit, deepgrow, or segmentation) to work! I'll attach some images to maybe help.   
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/ea48e4e1-cb5b-466e-864e-1fc1d74920b0)

I can't figure out why the "Model" section in auto Segmentation or SmartEdit/Deepgrow won't populate but I suspect that's where the problem lies. 

Here is the command I used to start the server: 
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/fff28669-ead9-4a4d-a03b-1ea4780333a5)

Thanks in advance for the help!

**Comment by coronaryangina001** [comment]
^the blacked out part is the directory for my CT scan that I'm using

**Comment by coronaryangina001** [comment]
I figured it out! I had to re-install my conda environment and then go in and re-install monailabel! Problem now is that when I run autosegmentation it just spits out a blob.

**Reply by diazandr3s** [reply]
Thanks for the update, @coronaryangina001. I'm glad you managed to figure this out.

> Problem now is that when I run autosegmentation it just spits out a blob.

Are you re-training the model? Can you provide more details of this task? How many regions/chambers do you want to segment? Do you have an annotated dataset?

**Reply by coronaryangina001** [reply]
Hey! Thanks for responding!! I am not re-training the model. I'm just trying to run what segmentation already has. I'm just trying to get the 4 chambers all separate, aorta, pulmonary artery, and the vena cavas! 

I have a non-annotated dataset I'm using.

Details for the task: essentially I'm just uploading my CT scan onto the server and I'm trying to get autosegmentation to run. I ran it using just my cpu and it just gave me a single blob for the aorta (no picture sadly). I have a RTX 3070 with 8GB of VRAM but I was getting CUDA memory errors with that. I tried  removing all of the excess segmentations that I didn't need from the segmenation config file but now I'm getting this:

[2024-02-27 19:58:28,883] [14452] [MainThread] [INFO] (monailabel.tasks.infer.basic_infer:504) - Inferer:: cuda => SlidingWindowInferer => {'roi_size': (96, 96, 96), 'sw_batch_size': 2, 'overlap': 0.4, 'mode': gaussian, 'sigma_scale': 0.125, 'padding_mode': 'replicate', 'cval': 0.0, 'sw_device': None, 'device': None, 'progress': False, 'cpu_thresh': None, 'buffer_steps': None, 'buffer_dim': -1, 'with_coord': False, 'roi_weight_map': None}
[2024-02-27 19:58:28,883] [14452] [MainThread] [INFO] (monailabel.tasks.infer.basic_infer:432) - Infer model path: C:\Users\name\apps\radiology\model\pretrained_segmentation.pt
[2024-02-27 19:58:28,985] [14452] [MainThread] [ERROR] (uvicorn.error:434) - Exception in ASGI application
Traceback (most recent call last):
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\uvicorn\protocols\http\h11_impl.py", line 429, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\uvicorn\middleware\proxy_headers.py", line 78, in __call__
    return await self.app(scope, receive, send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\fastapi\applications.py", line 276, in __call__
    await super().__call__(scope, receive, send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\applications.py", line 122, in __call__
    await self.middleware_stack(scope, receive, send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\middleware\errors.py", line 184, in __call__
    raise exc
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\middleware\errors.py", line 162, in __call__
    await self.app(scope, receive, _send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\middleware\cors.py", line 84, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\middleware\exceptions.py", line 79, in __call__
    raise exc
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\middleware\exceptions.py", line 68, in __call__
    await self.app(scope, receive, sender)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 21, in __call__
    raise e
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\fastapi\middleware\asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\routing.py", line 718, in __call__
    await route.handle(scope, receive, send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\routing.py", line 276, in handle
    await self.app(scope, receive, send)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\starlette\routing.py", line 66, in app
    response = await func(request)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\fastapi\routing.py", line 237, in app
    raw_response = await run_endpoint_function(
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\fastapi\routing.py", line 163, in run_endpoint_function
    return await dependant.call(**values)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\monailabel\endpoints\infer.py", line 206, in api_run_inference
    return run_inference(background_tasks, model, image, session_id, params, file, label, output)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\monailabel\endpoints\infer.py", line 172, in run_inference
    result = instance.infer(request)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\monailabel\interfaces\app.py", line 307, in infer
    result_file_name, result_json = task(request)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\monailabel\tasks\infer\basic_infer.py", line 307, in __call__
    data = self.run_inferer(data, device=device)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\monailabel\tasks\infer\basic_infer.py", line 506, in run_inferer
    network = self._get_network(device, data)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\monailabel\tasks\infer\basic_infer.py", line 480, in _get_network
    network.load_state_dict(model_state_dict, strict=self.load_strict)
  File "C:\Users\name\.conda\envs\monailabel-env\lib\site-packages\torch\nn\modules\module.py", line 2153, in load_state_dict
    raise RuntimeError('Error(s) in loading state_dict for {}:\n\t{}'.format(
RuntimeError: Error(s) in loading state_dict for SegResNet:
        size mismatch for conv_final.2.conv.weight: copying a param with shape torch.Size([26, 32, 1, 1, 1]) from checkpoint, the shape in current model is torch.Size([9, 32, 1, 1, 1]).
        size mismatch for conv_final.2.conv.bias: copying a param with shape torch.Size([26]) from checkpoint, the shape in current model is torch.Size([9])


Not quite sure what all of this means. I was trying to reduce the amount of memory being used all at once by removing the segment labels from the config file but I'm afraid that may have broken it lol. I feel like I'm extremely close to getting this to run and I can't wait to see what it can do!

**Reply by coronaryangina001** [reply]
I've read that you can reduce the batch in order to lower the amount of memory used by I can't figure that out unfortunately.

**Reply by coronaryangina001** [reply]
I got it to run on a CT with less slices, but here is what it gives me:
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/c9d39a48-1794-40b2-ae59-0303b9e0070d)

**Reply by diazandr3s** [reply]
Hi @coronaryangina001,

Unfortunately, we don't have a model for segmenting the 4 chambers, the aorta, the pulmonary artery, and the vena cava on CT volumes.

Here is a quick video showing how you could train a MONAI Label model from scratch: https://youtu.be/3HTh2dqZqew?si=eb_P87UO3U0oOjDE

You'll need an annotated dataset to get a first model and then trigger the active learning cycle.

You could use the "cardiac model" available in the MONAI Auto3DSeg module in Slicer to bootstrap the annotation process - correct the segments or add more segments to your volumes. 

For that, just download the latest Slicer and install MONAI Auto3D via the Extension Manager.

https://github.com/Project-MONAI/MONAILabel/assets/11991079/faeaeea6-c828-405e-a7be-689dda965bd5

You could also use the "quick" model with less resolution

Once you get an annotated dataset, you could use MONAI Label to improve the model accuracy on your task.

Let us know how that goes,

**Reply by coronaryangina001** [reply]
Oh okay! 

I just tried using the cardiac model in auto3dseg module but this is what it gave me: 

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/25bf1e44-e00a-4afc-83db-2543d357ca1f)

I'm not sure why that didn't work, does it have something to do with it not downloading the programs properly over on the left? I looked at the video you posted and it seemed as though yours downloaded more data then what mine did when you initialized it. 

Also, thank you for helping me with this. I really appreciate it!

**Reply by coronaryangina001** [reply]
It also looks like when it runs the inference yours has more bars doing computations than mine. Did I install something wrong?

**Reply by diazandr3s** [reply]
Hi @coronaryangina001,

Thanks for trying this out. 

Are these human CTs? Most of the models we have are for human CTs.

Please let us know,

**Reply by coronaryangina001** [reply]
Yup they are human! I noticed on your process is has 1/12 --> 12/12 on your screen when you run the program. On mine it just has 1/1. Do you think that might have something to do with it? I also tried it on the exact same CT you used in the video and it gave me a similar result. I've tried re-installing the plugin and still no luck.

**Reply by coronaryangina001** [reply]
I tried reinstalling slicer (version 5.6) and it's still showing the same thing.

**Reply by diazandr3s** [reply]
That's strange.

> I also tried it on the exact same CT you used in the video and it gave me a similar result. 

Similar to the result I've shown in the video or the print screen you've shared?

In any case, please download the Slicer version from today. I see you have the Slicer version from 3 days ago: 5.7.0.2024.02.26

Let us know,

**Reply by coronaryangina001** [reply]
Similar to what I showed. I'll download today's release of slicer and try it.

**Reply by pieper** [reply]
FWIW I tried with a windows preview version from a couple days ago (2024-02-27) and the Auto3DSeg extension worked great for my CT data.

**Reply by coronaryangina001** [reply]
Yeah it's definitely a problem on my end I just can't figure out what it is. I got the same result again:

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/e800e8fd-4bb6-47ad-ab07-30f104fa5339)
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/ca8f9aa0-e74b-4f4d-896e-f6bc43577461)

**Reply by pieper** [reply]
It's cut off, but that part about something not being installed correctly might be the issue.

**Reply by diazandr3s** [reply]
Can you share a print screen of the result on the sample data I've shown in the video?
Also, I could run inference on my PC if I have access to the volume you are using. Let me know

**Reply by coronaryangina001** [reply]
Will do, one sec

**Reply by coronaryangina001** [reply]
Whoa!!! It just worked!!! 
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/72ad2ee3-5667-4b0e-9a6d-f31c2b3578f7)
I just dont know why lol: This was the first time this happened: 
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/6bc3a72a-f21d-409a-aa61-11b2df1e89f1)
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/78a1805b-890e-488d-b6ff-36dcc65ac1f7)

**Reply by diazandr3s** [reply]
Did it work on your image as well?

**Reply by coronaryangina001** [reply]
Blast. It still has the same issue on my scan though: 

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/309bab83-0ffc-4d2e-b3e9-58dd60c45d80)

**Reply by diazandr3s** [reply]
Can you try with another image?

Also, please open the Python log  window (top right icon) and see what it prints?

**Reply by lassoan** [reply]
> Yup they are human

The images does not look anything like the usual human clinical CTs. Is this an infant?

Everything indicates that its size is half or third of an adult human (inference takes just a few iterations, logits nearly 0, result is low resolution, low quality). You will get better results if you scale up the physical size of your image by a factor of 2x or 3x (by doubling or tripling all 3 spacing values in Volumes module / Volume information section).

**Reply by coronaryangina001** [reply]
Yes this is a pediatric patient

**Reply by coronaryangina001** [reply]
It printed quite a lot:

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/cc72b724-874c-4085-88f8-f89abe179ae5)
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/61cd4500-3f19-4b46-a067-22556c83ef0e)
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/198e463b-c6c9-4bd7-b708-13a17d5c4192)
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/5b38fafd-1e6b-4368-a836-a1094af46616)
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/8de45dfb-e207-40a3-bc0b-4ee432cdb325)
There's more than that but hopefully this gives you an idea of what's going on?

**Reply by lassoan** [reply]
Thanks for confirming, this explains everything. You need to scale up the physical size before segmentation as I described above, because the processing starts with downsampling the image to fixed 1.5mm voxel size. Not much will be visible of a small baby at that low resolution and the segmentation algorithm may be confused by the huge size mismatch.

**Reply by coronaryangina001** [reply]
Thank you for helping!! So just multiple these numbers in "image spacing" by 3x should do it?
 
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/c8d9ad7b-1be2-4eee-9c47-c019adf53563)

**Reply by lassoan** [reply]
Those errors are related to how you are attempting to load the data set. It seems that you are not loading the DICOM module. DICOM images are loaded correctly by the DICOM module. If you have the volume in some series of TIFF/PNG/JPEG files then try to get the DICOM files.

**Reply by lassoan** [reply]
Yes, replace 0.313 by 0.939; 0.250 by 0.75.

**Reply by coronaryangina001** [reply]
IT WORKED!!!  
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/efea8fc4-d817-42a3-84c7-33e3c71307da)
 
YOU GUYS ARE AWESOME

**Reply by coronaryangina001** [reply]
Now how do I split the heart itself up by the four chambers lol!

**Reply by coronaryangina001** [reply]
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/39cebd12-250f-47be-9ea2-81a34a0016e4)
Is there a way to recenter the image so that I can go about editing the result?

**Reply by coronaryangina001** [reply]
Also, I gotta go, but you guys made my day. I've been working on this all week and to see this finally work has really been awesome. I hope you guys have a great day as well!!

**Reply by lassoan** [reply]
No need to recenter. Do you mean that the segmentation is not aligned with the underlying image?

**Reply by diazandr3s** [reply]
Thanks for the update, @coronaryangina001. Good to see it is working.

@lassoan thanks for your help!

With regards to this:

>  Now how do I split the heart itself up by the four chambers lol!

This video might help: https://www.youtube.com/watch?v=BJoIexIvtGo&t=3s

@lassoan do you think this is the fastest way?

**Reply by lassoan** [reply]
Yes, following the video is a good way to segment the heart. Maybe after segmenting 10-20 images, you can train an AI model and you just need to review and touch up the automatic segmentation results.

**Reply by diazandr3s** [reply]
Agree! @coronaryangina001 let me know once you get those images annotated. I can train a custom model for you.

**Reply by coronaryangina001** [reply]
That'd be great! Is that something I could feasibly do myself? Also, I noticed something, I went back in and tried to see if monailabel would still work from the server on my CPU, it seems as though when I ran the autosegmentation it somehow interfered with monai that I have installed through my anaconda powershell. It's resulted in the same problem I had before. I'm trying to reinstall the env again to see if I can restore it:

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/57a0ef07-bbd1-4e3c-9229-049430568fad)

**Reply by coronaryangina001** [reply]
this is what its throwing on startup, I re-installed pytorch w/ CUDA so it would see my graphics card: 

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/953649d2-bead-43b2-a514-b8dd73e73833)

**Reply by diazandr3s** [reply]
Hi @coronaryangina001,

> That'd be great! Is that something I could feasibly do myself?

Yes, you could train the model following this tutorial: https://github.com/Project-MONAI/tutorials/tree/main/auto3dseg#performance-benchmarking

Regarding MONAI Label, I see it started without issues. Have you tried fetching the next image? Which file format are you using to start the server?

**Reply by coronaryangina001** [reply]
I'll check that out as soon as I get the server issue fixed again. 

I have, I'm using the exact same file I've been using. Is it something to do with 3d slicer not recognizing the "segmentation" app running? 

I can retrieve my CT just fine using "next sample" 
![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/f6881c42-6176-42f8-ba59-b9a9f3bba4a8)
this won't let me select anything, last time this happened I installed a clean environment and that fixed it (which was the origin of this post) but now that isn't working.

**Reply by coronaryangina001** [reply]
Okay. I have a really really weird result going on. The "model" section is NOT letting me select anything in slicer 5.7.0 (latest release) but on the older slicer 5.2.2 it works. I'll post photos: 

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/dddf423b-c4fd-48be-82ab-c8bb68f512c8)

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/827780e6-591b-4320-8a35-92333098ebf6)

I think there is something wrong with the MonAi plugin maybe?

**Reply by coronaryangina001** [reply]
Okay. I think I got it figured out. If you have the old Nvidia AI Assissted Annotation downloaded then it causes that box to be greyed out. I uninstalled it and now it all works again:

![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/497b6e38-2318-4256-887b-1271d1b66fc8)


![image](https://github.com/Project-MONAI/MONAILabel/assets/161386765/68b31f7d-e24d-44d7-b327-a623b4c3d44d)

**Comment by diazandr3s** [comment]
> It's cut off, but that part about something not being installed correctly might be the issue.

That's true. I see it shows 0.0 seconds on the logits. The inference is not complete.
